ultrasonic/src/main/kotlin/org/moire/ultrasonic/fragment/SelectAlbumFragment.kt

```diff
@@ -0,0 +1,1180 @@
+            if (playAll && songCount > 0) {
+                playAll(
+                    arguments!!.getBoolean(Constants.INTENT_EXTRA_NAME_SHUFFLE, false),
+                    false
+                )
+            }
+        }
+
+        protected fun createHeader(
+            entries: List<MusicDirectory.Entry>,
+            name: CharSequence?,
+            songCount: Int
+        ): View? {
+            val coverArtView = header!!.findViewById<View>(R.id.select_album_art) as ImageView
+            val artworkSelection = random.nextInt(entries.size)
+            imageLoaderProvider.value.getImageLoader().loadImage(
+                coverArtView, entries[artworkSelection], false,
+                Util.getAlbumImageSize(context), false, true
+            )
+
+            val albumHeader = AlbumHeader.processEntries(context, entries)
+
+            val titleView = header!!.findViewById<View>(R.id.select_album_title) as TextView
+            titleView.text = name ?: getTitle(this@SelectAlbumFragment) // getActionBarSubtitle());
+
+            // Don't show a header if all entries are videos
+            if (albumHeader.isAllVideo) {
+                return null
+            }
+
+            val artistView = header!!.findViewById<TextView>(R.id.select_album_artist)
+            val artist: String
+
+            artist = if (albumHeader.artists.size == 1)
+                albumHeader.artists.iterator().next()
+            else if (albumHeader.grandParents.size == 1)
+                albumHeader.grandParents.iterator().next()
+            else
+                resources.getString(R.string.common_various_artists)
+
+            artistView.text = artist
+
+            val genreView = header!!.findViewById<TextView>(R.id.select_album_genre)
+            val genre: String
+
+            genre = if (albumHeader.genres.size == 1)
+                albumHeader.genres.iterator().next()
+            else
+                resources.getString(R.string.common_multiple_genres)
+
+            genreView.text = genre
+
+            val yearView = header!!.findViewById<TextView>(R.id.select_album_year)
+            val year: String
+
+            year = if (albumHeader.years.size == 1)
+                albumHeader.years.iterator().next().toString()
+            else
+                resources.getString(R.string.common_multiple_years)
+
+            yearView.text = year
+
+            val songCountView = header!!.findViewById<TextView>(R.id.select_album_song_count)
+            val songs = resources.getQuantityString(
+                R.plurals.select_album_n_songs, songCount,
+                songCount
+            )
+            songCountView.text = songs
+
+            val duration = Util.formatTotalDuration(albumHeader.totalDuration)
+
+            val durationView = header!!.findViewById<TextView>(R.id.select_album_duration)
+            durationView.text = duration
+
+            return header
+        }
+    }
+
+    companion object {
```

Comment:
Maybe these can be simple private members of the class, what do you think? Companion objects usually are necessary when they contain something public.

Reply:
yeah, that "companion" thing is new to me. I think that's how the Java->Kotlin conversion tools handle statics. I'm not sure why these things were static in the first place. I made them private and everything works just fine.

Reply:
I think the function was static because it was a pure function (didn't use instance properties, just its parameters) and those are usually marked as static in java and similar languages. Anyway, kotlin doesn't have static and the companion wasn't necessary (some use the companion in kotlin to store pure functions in it, but I'm not sure if that's a good practice - it looks ugly especially when you try to access them from java).